controllers/k8ssandra/medusa_reconciler.go

```diff
@@ -268,63 +271,76 @@ func (r *K8ssandraClusterReconciler) mergeStorageProperties(
-	if err := c.Get(ctx, medusaConfigKey, medusaConfig); err != nil {
-		logger.Error(err, fmt.Sprintf("could not get MedusaConfiguration %s/%s", medusaConfigNamespace, medusaConfigName))
-		return err
-	}
+		// there is nothing to reconcile if we're not using Medusa configuration reference
+		if medusaSpec == nil || medusaSpec.MedusaConfigurationRef.Name == "" {
+			logger.Info("MedusaConfigurationRef is not set, skipping bucket secret reconciliation")
+			return nil
+		}
 
-	// fetch the referenced medusa configuration's bucket secret
-	bucketSecretName := medusaConfig.Spec.StorageProperties.StorageSecretRef.Name
-	bucketSecret := &corev1.Secret{}
-	bucketSecretKey := types.NamespacedName{Namespace: medusaConfigNamespace, Name: bucketSecretName}
-	if err := c.Get(ctx, bucketSecretKey, bucketSecret); err != nil {
-		logger.Error(err, "could not get bucket Secret")
-		return err
-	}
+		// fetch the referenced configuration
+		medusaConfigName := medusaSpec.MedusaConfigurationRef.Name
+		medusaConfigNamespace := utils.FirstNonEmptyString(medusaSpec.MedusaConfigurationRef.Namespace, kc.Namespace)
+		medusaConfigKey := types.NamespacedName{Namespace: medusaConfigNamespace, Name: medusaConfigName}
+		medusaConfig := &medusaapi.MedusaConfiguration{}
+		if err := c.Get(ctx, medusaConfigKey, medusaConfig); err != nil {
+			logger.Error(err, fmt.Sprintf("could not get MedusaConfiguration %s/%s", medusaConfigNamespace, medusaConfigName))
+			return err
+		}
 
-	// write the secret into the namespace of the K8ssandraCluster
-	clusterBucketSecret := bucketSecret.DeepCopy()
-	clusterBucketSecret.ResourceVersion = ""
-	clusterBucketSecret.Name = fmt.Sprintf("%s-%s", kc.Name, bucketSecret.Name)
-	clusterBucketSecret.Namespace = kc.Namespace
-	labels.SetReplicatedBy(clusterBucketSecret, utils.GetKey(kc))
-	if err := c.Create(ctx, clusterBucketSecret); err != nil {
-		if !errors.IsAlreadyExists(err) {
-			logger.Error(err, fmt.Sprintf("failed to create cluster bucket secret %s", clusterBucketSecret))
+		// fetch the referenced medusa configuration's bucket secret
+		bucketSecretName := medusaConfig.Spec.StorageProperties.StorageSecretRef.Name
+		bucketSecret := &corev1.Secret{}
+		bucketSecretKey := types.NamespacedName{Namespace: medusaConfigNamespace, Name: bucketSecretName}
+		if err := c.Get(ctx, bucketSecretKey, bucketSecret); err != nil {
+			logger.Error(err, "could not get bucket Secret")
 			return err
 		}
-		// we already have the bucket secret, so continue to updating the cluster (it might have failed before)
-	}
 
-	return nil
+		// write the secret into the namespace of the K8ssandraCluster
+		clusterBucketSecret := bucketSecret.DeepCopy()
+		clusterBucketSecret.ResourceVersion = ""
+		clusterBucketSecret.Name = fmt.Sprintf("%s-%s", kc.Name, bucketSecret.Name)
+		clusterBucketSecret.Namespace = kc.Namespace
+		labels.SetReplicatedBy(clusterBucketSecret, utils.GetKey(kc))
+		if err := c.Create(ctx, clusterBucketSecret); err != nil {
```

Comment:
Ah, this comment is outdated. So originally, this PR was just to address #1266, however I've expanded it to embrace #1217 as well. The current set of commits has some more work in this direction.

> I assume you cannot use the ReplicatedSecret API to replicate one secret to multiple clusters, right? So that means we need to refresh our initial copies when the original secret is updated, and trigger a reconcile.

I don't quite follow this question - there will be two replicated secrets per cluster, one in the MedusaConfig ns and one in the K8ssandraCluster ns. This code path will only become active for legacy clusters which have non-namespace local MedusaConfigs.

See my comments throughout the code for more details on what code paths point where, this is a bit convoluted as we're supporting the legacy mode, but hopefully there are enough comments there that we can slice this functionality out easily when we've completed the deprecation cycle.